### DISCUSSION TITLE
Sending Custom Commands to Crazyflie Quadcopter Using Python cflib

### DISCUSSION THREAD

**Discussion by amitx64** [bitcraze/discussions#1229] [Q&A] [answered]
**Description:**
I'm working on a project involving controlling a Crazyflie quadcopter using Python and the Crazyflie Python library `cflib`. As part of this project, I need to send custom commands to the Crazyflie, such as starting, pausing, or stopping a drone, and more.

**Custom Commands:**
Here are the custom commands I'd like to implement:
```python
CMD_START = 0
CMD_PAUSE = 1
CMD_STOP = 2
CMD_STATUS = 3
CMD_RESTART = 4
CMD_TRIGGER_GCS_LIGHT_EFFECT = 5
CMD_ARM_OR_DISARM = 6
CMD_DEFINE_FENCE = 7
```

**Objective:**
I'm seeking advice and recommendations on how to implement these custom commands effectively. Specifically, I'd like assistance with:
- Developing the Python script to send these commands to the Crazyflie.
- Integrating these commands into the Crazyflie firmware.
- Ensuring the reliability and safety of the commands.

**Current Approach:**
I have already made progress with the Python script and have connected to the Crazyflie using the Crazyflie Python library. However, I'm unsure about the best way to send these custom commands and integrate them into the Crazyflie firmware.

**Request for Feedback:**
- What would be the best way to send custom commands to the Crazyflie from a Python script?
- How can I integrate these commands into the Crazyflie firmware effectively?
- Are there any safety considerations I should keep in mind when sending commands to the Crazyflie?

Any advice, suggestions, or resources would be greatly appreciated. Thank you in advance for your help!

**Comment by gemenerik** [comment, marked as answer]
I would recommend writing an app on the crazyflie-firmware side and using the app channel to communicate with it from the cflib. A good starting point may be [this example](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/app_appchannel_test) and the documentation on the [app layer](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/). If you are concerned about safety read you can e.g. connect the cflib `send_emergency_stop()` fn to a keypress. For more information about safety you can read [this (slightly outdated) blogpost](https://www.bitcraze.io/2023/04/safety-and-the-brushless/)

**Reply by amitx64** [reply]
@gemenerik I wanted to extend my sincere gratitude for taking the time to provide a solution. Your solution was insightful and well-explained, making it easy for me to understand and implement. Once again, Thank You!
